File: src/storage/database/shared/schema.ts
import { pgTable, serial, text, varchar, integer, timestamp, date, boolean } from "drizzle-orm/pg-core"
import { sql } from "drizzle-orm"

// ============ 系统表 ============

export const healthCheck = pgTable("health_check", {
  id: serial().primaryKey(),
  updatedAt: timestamp("updated_at", { withTimezone: true, mode: 'string' }).defaultNow(),
});

// ============ 板块元信息 ============

export const boardMeta = pgTable("board_meta", {
  boardId: text("board_id").primaryKey(),
  updated: text("updated"),
  source: text("source"),
  module: text("module"),
});

// ============ 板块状态 ============

export const boardStatus = pgTable("board_status", {
  boardId: text("board_id"),
  lastCrawledAt: text("last_crawled_at"),
  newItemsCount: text("new_items_count"),
  totalSources: text("total_sources"),
  errorSources: text("error_sources"),
  lastMessage: text("last_message"),
  rocketIntro: text("rocket_intro"),
});

// ============ 可回收火箭 ============

export const rocketCompanies = pgTable("rocket_companies", {
  id: integer("id").primaryKey(),
  rocket: text("rocket"),
  company: text("company"),
  country: text("country"),
  fuel: text("fuel"),
  diameter: text("diameter"),
  thrust: text("thrust"),
  leo: text("leo"),
  recovery: text("recovery"),
  status: text("status"),
  key: text("key"),
  sortOrder: text("sort_order"),
});

export const rocketLaunchTimeline = pgTable("rocket_launch_timeline", {
  timelineId: text("timeline_id").primaryKey(),
  rocketId: text("rocket_id"),
  missionName: text("mission_name"),
  launchTime: text("launch_time"),
  launchSite: text("launch_site"),
  payload: text("payload"),
  outcome: text("outcome"),
  reuseStatus: text("reuse_status"),
  briefDesc: text("brief_desc"),
  relatedNewsIds: json("related_news_ids"),
  createTime: text("create_time"),
  updateTime: text("update_time"),
});

// ============ 中美登月 ============

export const moonHighlights = pgTable("moon_highlights", {
  id: integer("id").primaryKey(),
  num: text("num"),
  label: text("label"),
  color: text("color"),
  sortOrder: text("sort_order"),
});

export const moonComparison = pgTable("moon_comparison", {
  id: serial().primaryKey(),
  aspect: varchar("aspect"),
  chinaValue: text("china_value"),
  usaValue: text("usa_value"),
  notes: text("notes"),
  createdAt: timestamp("created_at", { mode: 'string' }).defaultNow(),
});

// ============ 中国半导体 ============

export const semiconductorHighlights = pgTable("semiconductor_highlights", {
  id: integer("id").primaryKey().generatedAlwaysAsIdentity(),
  num: text("num").notNull(),
  label: text("label").notNull(),
  color: text("color"),
  sortOrder: integer("sort_order").default(0),
});

export const semiconductorTabHighlights = pgTable("semiconductor_tab_highlights", {
  id: integer("id").primaryKey().generatedAlwaysAsIdentity(),
  tabId: text("tab_id").notNull(),
  num: text("num").notNull(),
  label: text("label").notNull(),
  color: text("color"),
  sortOrder: integer("sort_order").default(0),
});

export const semiconductorTabProgress = pgTable("semiconductor_tab_progress", {
  id: integer("id").primaryKey().generatedAlwaysAsIdentity(),
  tabId: text("tab_id").notNull(),
  year: text("year").notNull(),
  value: text("value"),
  label: text("label"),
  cls: text("cls"),
  sortOrder: integer("sort_order").default(0),
});

export const semiconductorTechnologies = pgTable("semiconductor_technologies", {
  id: serial().primaryKey(),
  name: varchar("name").notNull(),
  category: varchar("category"),
  description: text("description"),
  status: varchar("status"),
  createdAt: timestamp("created_at", { mode: 'string' }).defaultNow(),
});

export const semiconductorTimeline = pgTable("semiconductor_timeline", {
  id: serial().primaryKey(),
  eventDate: date("event_date"),
  company: varchar("company"),
  eventType: varchar("event_type"),
  description: text("description"),
  createdAt: timestamp("created_at", { mode: 'string' }).defaultNow(),
});

// ============ 中国科技 AI ============

export const chinaTechHighlights = pgTable("china_tech_highlights", {
  id: integer("id").primaryKey().generatedAlwaysAsIdentity(),
  num: text("num").notNull(),
  label: text("label").notNull(),
  color: text("color"),
  sortOrder: integer("sort_order").default(0),
});

export const chinaTechLlm = pgTable("china_tech_llm", {
  id: integer("id").primaryKey().generatedAlwaysAsIdentity(),
  model: text("model").notNull(),
  company: text("company").notNull(),
  params: text("params"),
  contextWindow: text("context_window"),
  coding: text("coding"),
  math: text("math"),
  arena: text("arena"),
  opensource: text("opensource"),
  price: text("price"),
  hiFields: text("hi_fields"),
  sortOrder: integer("sort_order").default(0),
});

export const chinaTechTimeline = pgTable("china_tech_timeline", {
  id: serial().primaryKey(),
  eventDate: date("event_date"),
  company: varchar("company"),
  eventType: varchar("event_type"),
  description: text("description"),
  createdAt: timestamp("created_at", { mode: 'string' }).defaultNow(),
});

// ============ 中国大工程 ============

export const megaProjects = pgTable("mega_projects", {
  id: integer("id").primaryKey(),
  tabId: text("tab_id"),
  emoji: text("emoji"),
  projectName: text("project_name"),
  targetId: text("target_id"),
  status: text("status"),
  statusClass: text("status_class"),
  sortOrder: text("sort_order"),
});

export const megaProjectHighlights = pgTable("mega_project_highlights", {
  id: integer("id").primaryKey().generatedAlwaysAsIdentity(),
  num: text("num").notNull(),
  label: text("label").notNull(),
  color: text("color"),
  sortOrder: integer("sort_order").default(0),
});

export const megaProjectMilestones = pgTable("mega_project_milestones", {
  id: integer("id").primaryKey().generatedAlwaysAsIdentity(),
  projectId: integer("project_id").notNull(),
  marker: text("marker"),
  eventDate: text("event_date"),
  badge: text("badge"),
  badgeClass: text("badge_class"),
  title: text("title"),
  sortOrder: integer("sort_order").default(0),
});

// ============ 可控核聚变 ============

export const fusionHighlights = pgTable("fusion_highlights", {
  id: integer("id").primaryKey().generatedAlwaysAsIdentity(),
  num: text("num").notNull(),
  label: text("label").notNull(),
  color: text("color"),
  sortOrder: integer("sort_order").default(0),
});

export const fusionTimeline = pgTable("fusion_timeline", {
  id: integer("id").primaryKey(),
  region: text("region"),
  regionLabel: text("region_label"),
  eventDate: text("event_date"),
  title: text("title"),
  description: text("description"),
  color: text("color"),
  sortOrder: text("sort_order"),
});

// ============ 科技资本 ============

export const financeGrids = pgTable("finance_grids", {
  id: integer("id").primaryKey().generatedAlwaysAsIdentity(),
  section: text("section").notNull(),
  key: text("key").notNull(),
  value: text("value").notNull(),
  sortOrder: integer("sort_order").default(0),
});

export const financeHighlights = pgTable("finance_highlights", {
  id: integer("id").primaryKey().generatedAlwaysAsIdentity(),
  section: text("section").notNull(),
  label: text("label").notNull(),
  num: text("num").notNull(),
  sub: text("sub"),
  color: text("color"),
  sortOrder: integer("sort_order").default(0),
});

export const financeSections = pgTable("finance_sections", {
  id: integer("id").primaryKey().generatedAlwaysAsIdentity(),
  section: text("section").notNull(),
  tag: text("tag"),
  name: text("name"),
  en: text("en"),
  description: text("description"),
  sortOrder: integer("sort_order").default(0),
});

// ============ 新闻数据（统一原始文章表）============

export const rawArticles = pgTable("raw_articles", {
  newsId: text("news_id").primaryKey(),           // 唯一哈希ID（URL哈希/UUID）
  sourceName: text("source_name"),                 // 来源媒体名称
  sourceUrl: text("source_url"),                   // 原文链接（去重依据）
  crawlTime: timestamp("crawl_time", { mode: 'string' }).defaultNow(),  // 抓取时间
  publishTime: timestamp("publish_time", { mode: 'string' }),  // 新闻发布时间
  title: text("title"),                            // 新闻标题
  rawContent: text("raw_content"),                 // 原始正文文本
  summary: text("summary"),                        // AI生成摘要
  coverImage: text("cover_image"),                 // 封面图片链接
  images: text("images"),                          // 文中图片数组（JSON）
  tags: text("tags"),                              // AI自动标签（JSON数组）
  category: text("category"),                      // 一级分类（航天/核聚变/半导体等）
  hotScore: integer("hot_score").default(0),       // AI热度分值 0-100
  sentiment: text("sentiment"),                    // 情感倾向 positive/neutral/negative
  eventGroupId: text("event_group_id"),            // 事件聚类ID
  language: text("language").default("en"),        // 语言 zh/en
  status: text("status").default("pending"),       // 生命周期 pending/online/block
});

export const crawlLogs = pgTable("crawl_logs", {
  id: serial().primaryKey(),
  boardId: varchar("board_id"),
  status: varchar("status"),
  sourceName: varchar("source_name"),
  itemsCount: integer("items_count").default(0),
  errorMessage: text("error_message"),
  startedAt: timestamp("started_at", { mode: 'string' }),
  finishedAt: timestamp("finished_at", { mode: 'string' }),
  createdAt: timestamp("created_at", { mode: 'string' }).defaultNow(),
});


// ============ 最新要闻（首页）============

export const latestNews = pgTable("latest_news", {
  id: serial().primaryKey(),
  title: text("title").notNull(),
  summary: text("summary").notNull(),
  source: text("source"),
  board: text("board").notNull(),
  boardLabel: text("board_label"),
  link: text("link"),
  publishDate: date("publish_date"),
  sortOrder: integer("sort_order").default(0),
  isActive: boolean("is_active").default(true),
  createdAt: timestamp("created_at", { withTimezone: true, mode: 'string' }).defaultNow(),
  updatedAt: timestamp("updated_at", { withTimezone: true, mode: 'string' }).defaultNow(),
});

File: backend/app/db/local/schema.py
# -*- coding: utf-8 -*-
"""
钛星本地 SQLite 建表定义（schema）

权威来源（按协调者指令）：
1. src/storage/database/shared/schema.ts（drizzle pg 定义，26 张表）
2. docs/sql/ai_extract_logs.sql（ai_extract_logs 表）
3. backend/app/db/*.py 现有代码推断（rocket_timeline 表，以及 raw_articles 的补充列）

类型约定（本地 SQLite 映射）：
- TEXT   ：文本 / varchar / date / timestamp（ISO 8601 字符串，与现有代码 datetime.now().isoformat() 一致）
- INTEGER：整数（含 id 主键、hot_score、sort_order 等）
- REAL   ：浮点（当前 schema 无 REAL 列，预留）
- JSON   ：以 TEXT 存储，写入 json.dumps、读取 json.loads（如 rocket_launch_timeline.related_news_ids）
- BOOLEAN：以 INTEGER 0/1 存储，读取转回 bool（如 latest_news.is_active、rocket_timeline.done）

所有建表/建索引语句幂等（CREATE TABLE/INDEX IF NOT EXISTS）。
"""

from typing import Dict, List, Tuple


def col(name: str, typ: str = "TEXT", constraints: str = "", **flags) -> Tuple:
    """列定义：(name, sqlite_type, constraints, flags)。flags 支持 json=True / bool=True。"""
    return (name, typ, constraints, flags)


# ========================================================================
# 表定义（顺序即建表顺序）
# ========================================================================
TABLES: Dict[str, List[Tuple]] = {
    # ---------------- 系统表（schema.ts） ----------------
    "health_check": [
        col("id", "INTEGER", "PRIMARY KEY AUTOINCREMENT"),
        col("updated_at", "TEXT", "DEFAULT (datetime('now'))"),
    ],
    "board_meta": [
        col("board_id", "TEXT", "PRIMARY KEY"),
        col("updated", "TEXT"),
        col("source", "TEXT"),
        col("module", "TEXT"),
    ],
    "board_status": [
        col("id", "INTEGER", "PRIMARY KEY AUTOINCREMENT"),
        col("board_id", "TEXT"),
        col("last_crawled_at", "TEXT"),
        col("new_items_count", "TEXT"),
        col("total_sources", "TEXT"),
        col("error_sources", "TEXT"),
        col("last_message", "TEXT"),
        col("rocket_intro", "TEXT"),
        col("rocket_next_intro", "TEXT"),
    ],

    # ---------------- 可回收火箭（schema.ts） ----------------
    "rocket_companies": [
        col("id", "INTEGER", "PRIMARY KEY"),
        col("rocket", "TEXT"),
        col("company", "TEXT"),
        col("country", "TEXT"),
        col("fuel", "TEXT"),
        col("diameter", "TEXT"),
        col("thrust", "TEXT"),
        col("leo", "TEXT"),
        col("recovery", "TEXT"),
        col("status", "TEXT"),
        col("key", "TEXT"),
        col("sort_order", "TEXT"),
    ],
    "rocket_launch_timeline": [
        col("timeline_id", "TEXT", "PRIMARY KEY"),
        col("rocket_id", "TEXT"),
        col("mission_name", "TEXT"),
        col("launch_time", "TEXT"),
        col("launch_site", "TEXT"),
        col("payload", "TEXT"),
        col("outcome", "TEXT"),
        col("reuse_status", "TEXT"),
        col("brief_desc", "TEXT"),
        col("related_news_ids", "TEXT", "", json=True),
        col("create_time", "TEXT"),
        col("update_time", "TEXT"),
    ],
    # 代码推断表：get_rocket_timeline() 查询 "rocket_timeline"，行按 period 字段（h2/2027）过滤，
    # 行结构与 data/rocket.json 的 timeline_h2/timeline_2027 条目一致（date/title/desc/color/badge/done）。
    "rocket_timeline": [
        col("id", "INTEGER", "PRIMARY KEY AUTOINCREMENT"),
        col("period", "TEXT"),
        col("date", "TEXT"),
        col("title", "TEXT"),
        col("desc", "TEXT"),
        col("color", "TEXT"),
        col("badge", "TEXT"),
        col("done", "INTEGER", "DEFAULT 0", bool=True),
        col("sort_order", "INTEGER", "DEFAULT 0"),
        col("created_at", "TEXT", "DEFAULT (datetime('now'))"),
    ],

    # ---------------- 中美登月（schema.ts） ----------------
    "moon_highlights": [
        col("id", "INTEGER", "PRIMARY KEY"),
        col("num", "TEXT"),
        col("label", "TEXT"),
        col("color", "TEXT"),
        col("sort_order", "TEXT"),
    ],
    "moon_comparison": [
        col("id", "INTEGER", "PRIMARY KEY AUTOINCREMENT"),
        col("aspect", "TEXT"),
        col("china_value", "TEXT"),
        col("usa_value", "TEXT"),
        col("notes", "TEXT"),
        col("created_at", "TEXT", "DEFAULT (datetime('now'))"),
    ],

    # ---------------- 中国半导体（schema.ts） ----------------
    "semiconductor_highlights": [
        col("id", "INTEGER", "PRIMARY KEY AUTOINCREMENT"),
        col("num", "TEXT", "NOT NULL"),
        col("label", "TEXT", "NOT NULL"),
        col("color", "TEXT"),
        col("sort_order", "INTEGER", "DEFAULT 0"),
    ],
    "semiconductor_tab_highlights": [
        col("id", "INTEGER", "PRIMARY KEY AUTOINCREMENT"),
        col("tab_id", "TEXT", "NOT NULL"),
        col("num", "TEXT", "NOT NULL"),
        col("label", "TEXT", "NOT NULL"),
        col("color", "TEXT"),
        col("sort_order", "INTEGER", "DEFAULT 0"),
    ],
    "semiconductor_tab_progress": [
        col("id", "INTEGER", "PRIMARY KEY AUTOINCREMENT"),
        col("tab_id", "TEXT", "NOT NULL"),
        col("year", "TEXT", "NOT NULL"),
        col("value", "TEXT"),
        col("label", "TEXT"),
        col("cls", "TEXT"),
        col("sort_order", "INTEGER", "DEFAULT 0"),
    ],
    "semiconductor_technologies": [
        col("id", "INTEGER", "PRIMARY KEY AUTOINCREMENT"),
        col("name", "TEXT", "NOT NULL"),
        col("category", "TEXT"),
        col("description", "TEXT"),
        col("status", "TEXT"),
        col("created_at", "TEXT", "DEFAULT (datetime('now'))"),
    ],
    "semiconductor_timeline": [
        col("id", "INTEGER", "PRIMARY KEY AUTOINCREMENT"),
        col("event_date", "TEXT"),
        col("company", "TEXT"),
        col("event_type", "TEXT"),
        col("description", "TEXT"),
        col("created_at", "TEXT", "DEFAULT (datetime('now'))"),
    ],

    # ---------------- 中国科技AI（schema.ts） ----------------
    "china_tech_highlights": [
        col("id", "INTEGER", "PRIMARY KEY AUTOINCREMENT"),
        col("num", "TEXT", "NOT NULL"),
        col("label", "TEXT", "NOT NULL"),
        col("color", "TEXT"),
        col("sort_order", "INTEGER", "DEFAULT 0"),
    ],
    "china_tech_llm": [
        col("id", "INTEGER", "PRIMARY KEY AUTOINCREMENT"),
        col("model", "TEXT", "NOT NULL"),
        col("company", "TEXT", "NOT NULL"),
        col("params", "TEXT"),
        col("context_window", "TEXT"),
        col("coding", "TEXT"),
        col("math", "TEXT"),
        col("arena", "TEXT"),
        col("opensource", "TEXT"),
        col("price", "TEXT"),
        col("hi_fields", "TEXT"),
        col("sort_order", "INTEGER", "DEFAULT 0"),
    ],
    "china_tech_timeline": [
        col("id", "INTEGER", "PRIMARY KEY AUTOINCREMENT"),
        col("event_date", "TEXT"),
        col("company", "TEXT"),
        col("event_type", "TEXT"),
        col("description", "TEXT"),
        col("created_at", "TEXT", "DEFAULT (datetime('now'))"),
    ],

    # ---------------- 中国大工程（schema.ts） ----------------
    "mega_projects": [
        col("id", "INTEGER", "PRIMARY KEY"),
        col("tab_id", "TEXT"),
        col("emoji", "TEXT"),
        col("project_name", "TEXT"),
        col("target_id", "TEXT"),
        col("status", "TEXT"),
        col("status_class", "TEXT"),
        col("sort_order", "TEXT"),
    ],
    "mega_project_highlights": [
        col("id", "INTEGER", "PRIMARY KEY AUTOINCREMENT"),
        col("num", "TEXT", "NOT NULL"),
        col("label", "TEXT", "NOT NULL"),
        col("color", "TEXT"),
        col("sort_order", "INTEGER", "DEFAULT 0"),
    ],
    "mega_project_milestones": [
        col("id", "INTEGER", "PRIMARY KEY AUTOINCREMENT"),
        col("project_id", "INTEGER", "NOT NULL"),
        col("marker", "TEXT"),
        col("event_date", "TEXT"),
        col("badge", "TEXT"),
        col("badge_class", "TEXT"),
        col("title", "TEXT"),
        col("sort_order", "INTEGER", "DEFAULT 0"),
    ],

    # ---------------- 可控核聚变（schema.ts） ----------------
    "fusion_highlights": [
        col("id", "INTEGER", "PRIMARY KEY AUTOINCREMENT"),
        col("num", "TEXT", "NOT NULL"),
        col("label", "TEXT", "NOT NULL"),
        col("color", "TEXT"),
        col("sort_order", "INTEGER", "DEFAULT 0"),
    ],
    "fusion_timeline": [
        col("id", "INTEGER", "PRIMARY KEY"),
        col("region", "TEXT"),
        col("region_label", "TEXT"),
        col("event_date", "TEXT"),
        col("title", "TEXT"),
        col("description", "TEXT"),
        col("color", "TEXT"),
        col("sort_order", "TEXT"),
    ],

    # ---------------- 科技资本（schema.ts） ----------------
    "finance_grids": [
        col("id", "INTEGER", "PRIMARY KEY AUTOINCREMENT"),
        col("section", "TEXT", "NOT NULL"),
        col("key", "TEXT", "NOT NULL"),
        col("value", "TEXT", "NOT NULL"),
        col("sort_order", "INTEGER", "DEFAULT 0"),
    ],
    "finance_highlights": [
        col("id", "INTEGER", "PRIMARY KEY AUTOINCREMENT"),
        col("section", "TEXT", "NOT NULL"),
        col("label", "TEXT", "NOT NULL"),
        col("num", "TEXT", "NOT NULL"),
        col("sub", "TEXT"),
        col("color", "TEXT"),
        col("sort_order", "INTEGER", "DEFAULT 0"),
    ],
    "finance_sections": [
        col("id", "INTEGER", "PRIMARY KEY AUTOINCREMENT"),
        col("section", "TEXT", "NOT NULL"),
        col("tag", "TEXT"),
        col("name", "TEXT"),
        col("en", "TEXT"),
        col("description", "TEXT"),
        col("sort_order", "INTEGER", "DEFAULT 0"),
    ],

    # ---------------- 新闻数据（schema.ts，raw_articles 补充代码推断列） ----------------
    # schema.ts 列：news_id/source_name/source_url/crawl_time/publish_time/title/raw_content/summary/
    #               cover_image/images/tags/category/hot_score/sentiment/event_group_id/language/status
    # 代码推断补充列（backend/app/db/articles.py 等写入）：
    #   id（select("id") 计数用）、dedup_key、board_id、source、url、date、raw_json、created_at
    # 注：news_id 在 schema.ts 为 PK；为兼容 legacy upsert_article（不带 news_id 写入）降级为可空 UNIQUE。
    "raw_articles": [
        col("id", "INTEGER", "PRIMARY KEY AUTOINCREMENT"),
        col("news_id", "TEXT", "UNIQUE"),
        col("source_name", "TEXT"),
        col("source_url", "TEXT"),
        col("crawl_time", "TEXT"),
        col("publish_time", "TEXT"),
        col("title", "TEXT"),
        col("raw_content", "TEXT"),
        col("summary", "TEXT"),
        col("cover_image", "TEXT"),
        col("images", "TEXT"),
        col("tags", "TEXT"),
        col("category", "TEXT"),
        col("hot_score", "INTEGER", "DEFAULT 0"),
        col("sentiment", "TEXT"),
        col("event_group_id", "TEXT"),
        col("language", "TEXT", "DEFAULT 'en'"),
        col("status", "TEXT", "DEFAULT 'pending'"),
        col("dedup_key", "TEXT", "UNIQUE"),
        col("board_id", "TEXT"),
        col("source", "TEXT"),
        col("url", "TEXT"),
        col("date", "TEXT"),
        col("raw_json", "TEXT"),
        col("created_at", "TEXT", "DEFAULT (datetime('now'))"),
    ],
    "crawl_logs": [
        col("id", "INTEGER", "PRIMARY KEY AUTOINCREMENT"),
        col("board_id", "TEXT"),
        col("status", "TEXT"),
        col("source_name", "TEXT"),
        col("items_count", "INTEGER", "DEFAULT 0"),
        col("error_message", "TEXT"),
        col("started_at", "TEXT"),
        col("finished_at", "TEXT"),
        col("created_at", "TEXT", "DEFAULT (datetime('now'))"),
    ],
    "latest_news": [
        col("id", "INTEGER", "PRIMARY KEY AUTOINCREMENT"),
        col("title", "TEXT", "NOT NULL"),
        col("summary", "TEXT", "NOT NULL"),
        col("source", "TEXT"),
        col("board", "TEXT", "NOT NULL"),
        col("board_label", "TEXT"),
        col("link", "TEXT"),
        col("publish_date", "TEXT"),
        col("sort_order", "INTEGER", "DEFAULT 0"),
        col("is_active", "INTEGER", "DEFAULT 1", bool=True),
        col("created_at", "TEXT", "DEFAULT (datetime('now'))"),
        col("updated_at", "TEXT", "DEFAULT (datetime('now'))"),
    ],

    # ---------------- AI 提取日志（docs/sql/ai_extract_logs.sql） ----------------
    "ai_extract_logs": [
        col("id", "INTEGER", "PRIMARY KEY AUTOINCREMENT"),
        col("category", "TEXT"),
        col("limit", "INTEGER"),
        col("total", "INTEGER"),
        col("inserted", "INTEGER"),
        col("failed", "INTEGER"),
        col("status", "TEXT"),
        col("message", "TEXT"),
        col("created_at", "TEXT", "DEFAULT (datetime('now'))"),
    ],
}

# (table, index_name, columns, unique)
INDEXES: List[tuple] = [
    ("raw_articles", "idx_raw_articles_status", ["status"], False),
    ("raw_articles", "idx_raw_articles_category", ["category"], False),
    ("raw_articles", "idx_raw_articles_crawl_time", ["crawl_time"], False),
    ("raw_articles", "idx_raw_articles_created_at", ["created_at"], False),
    ("crawl_logs", "idx_crawl_logs_created_at", ["created_at"], False),
    ("latest_news", "idx_latest_news_is_active", ["is_active"], False),
    ("latest_news", "idx_latest_news_sort_order", ["sort_order"], False),
    ("ai_extract_logs", "idx_ai_extract_logs_created_at", ["created_at"], False),
    ("board_status", "idx_board_status_board_id", ["board_id"], False),
    ("rocket_launch_timeline", "idx_rocket_launch_timeline_launch_time", ["launch_time"], False),
]

ALL_TABLES: List[str] = list(TABLES.keys())


def build_create_sql(name: str) -> str:
    """生成单表幂等建表 SQL"""
    if name not in TABLES:
        raise KeyError(f"[local-db] schema 未定义表: {name!r}")
    lines = []
    for cname, ctype, constraints, _flags in TABLES[name]:
        line = f'  "{cname}" {ctype}'.rstrip()
        if constraints:
            line += " " + constraints
        lines.append(line.rstrip())
    return f'CREATE TABLE IF NOT EXISTS "{name}" (\n' + ",\n".join(lines) + "\n);"


def build_index_sql(table: str, index_name: str, columns: List[str], unique: bool = False) -> str:
    cols = ", ".join(f'"{c}"' for c in columns)
    uniq = "UNIQUE " if unique else ""
    return f'CREATE {uniq}INDEX IF NOT EXISTS "{index_name}" ON "{table}" ({cols});'


def get_columns(name: str) -> List[str]:
    return [c[0] for c in TABLES[name]]


def get_column_map(name: str) -> Dict[str, Dict]:
    """列名 -> {type, json, bool}"""
    return {c[0]: {"type": c[1], "json": bool(c[3].get("json")), "bool": bool(c[3].get("bool"))} for c in TABLES[name]}


def init_db(conn) -> None:
    """执行全部建表/建索引 SQL（幂等）"""
    for name in TABLES:
        conn.execute(build_create_sql(name))
    for table, index_name, columns, unique in INDEXES:
        conn.execute(build_index_sql(table, index_name, columns, unique))
    conn.commit()

